source/adapters/opencl/kernel.hpp

```diff
@@ -0,0 +1,94 @@
+//===--------- kernel.hpp - OpenCL Adapter ---------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include "common.hpp"
+#include "context.hpp"
+
+#include <vector>
+
+struct ur_kernel_handle_t_ {
+  using native_type = cl_kernel;
+  native_type Kernel;
+  ur_program_handle_t Program;
+  ur_context_handle_t Context;
+  std::atomic<uint32_t> RefCount = 0;
+
+  ur_kernel_handle_t_(native_type Kernel, ur_program_handle_t Program,
+                      ur_context_handle_t Context)
+      : Kernel(Kernel), Program(Program), Context(Context) {
+    RefCount = 1;
+    if (Program) {
+      urProgramRetain(Program);
+    }
+    if (Context) {
+      urContextRetain(Context);
+    }
+  }
+
+  ~ur_kernel_handle_t_() {
+    clReleaseKernel(Kernel);
+    if (Program) {
+      urProgramRelease(Program);
+    }
+    if (Context) {
+      urContextRelease(Context);
+    }
+  }
+
+  uint32_t incrementReferenceCount() noexcept { return ++RefCount; }
+
+  uint32_t decrementReferenceCount() noexcept { return --RefCount; }
+
+  uint32_t getReferenceCount() const noexcept { return RefCount; }
+
+  static ur_result_t makeWithNative(native_type NativeKernel,
+                                    ur_program_handle_t Program,
+                                    ur_context_handle_t Context,
+                                    ur_kernel_handle_t &Kernel) {
+    try {
+      auto URKernel =
+          std::make_unique<ur_kernel_handle_t_>(NativeKernel, Program, Context);
+      if (!Program) {
```

Comment:
the program isn't optional in `KernelCreateWithNativeHandle`, do we fail an interop test if we remove this section the same way we do with `ContextCreateWithNative` and the device list?

Reply:
We did, but I moved all this logic to intel/llvm side and will stick with that and see the tests results, if any failure happens i will try to keep that logic with the fix there.